pkg/ingester/metrics.go

```diff
@@ -0,0 +1,186 @@
+package ingester
+
+import (
+	"sync"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+type ingesterMetrics struct {
+	flushQueueLength    prometheus.Gauge
+	ingestedSamples     prometheus.Counter
+	ingestedSamplesFail prometheus.Counter
+	queries             prometheus.Counter
+	queriedSamples      prometheus.Histogram
+	queriedSeries       prometheus.Histogram
+	queriedChunks       prometheus.Histogram
+}
+
+func newIngesterMetrics(r prometheus.Registerer) *ingesterMetrics {
+	m := &ingesterMetrics{
+		flushQueueLength: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_ingester_flush_queue_length",
+			Help: "The total number of series pending in the flush queue.",
+		}),
+		ingestedSamples: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ingester_ingested_samples_total",
+			Help: "The total number of samples ingested.",
+		}),
+		ingestedSamplesFail: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ingester_ingested_samples_failures_total",
+			Help: "The total number of samples that errored on ingestion.",
+		}),
+		queries: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ingester_queries_total",
+			Help: "The total number of queries the ingester has handled.",
+		}),
+		queriedSamples: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name: "cortex_ingester_queried_samples",
+			Help: "The total number of samples returned from queries.",
+			// Could easily return 10m samples per query - 10*(8^(8-1)) = 20.9m.
+			Buckets: prometheus.ExponentialBuckets(10, 8, 8),
+		}),
+		queriedSeries: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name: "cortex_ingester_queried_series",
+			Help: "The total number of series returned from queries.",
+			// A reasonable upper bound is around 100k - 10*(8^(6-1)) = 327k.
+			Buckets: prometheus.ExponentialBuckets(10, 8, 6),
+		}),
+		queriedChunks: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name: "cortex_ingester_queried_chunks",
+			Help: "The total number of chunks returned from queries.",
+			// A small number of chunks per series - 10*(8^(7-1)) = 2.6m.
+			Buckets: prometheus.ExponentialBuckets(10, 8, 7),
+		}),
+	}
+
+	if r != nil {
+		r.MustRegister(
+			m.flushQueueLength,
+			m.ingestedSamples,
+			m.ingestedSamplesFail,
+			m.queries,
+			m.queriedSamples,
+			m.queriedSeries,
+			m.queriedChunks,
+		)
+	}
+
+	return m
+}
+
+// TSDB shipper metrics
```

Comment:
maybe just add a further comment explaining this is a multi-tenant set of metrics that get aggregated into a single set of metrics?

Reply:
I've updated the comment. Is that what you had in mind?

Reply:
Looks great, thanks!

Reply:
Thank you for your review!